### ISSUE TITLE
Epic: ECE and Subscription products

### ISSUE THREAD

**Issue by bborman22** [Automattic/woocommerce-payments#9983]
This epic is to help us track our progress and compatibility with the mix of combinations available to subscription products covering: variable subscriptions, virtual products, free trials, and sign up fees. @asumaran had done some great work and generated the following chart to show our availability currently and we've been using that to map out our next steps. I'm finally creating this issue to keep any remainder of that discussion and planning in one place going forward.

> **Updated 2026-03-26** — Matrix below reflects the current state after tokenized ECE became the default ([#10016](<https://github.com/Automattic/woocommerce-payments/issues/10016>), merged Jan 2025) and subsequent subscription fixes (Automattic/woocommerce-payments#11340, Automattic/woocommerce-payments#11425). The old non-tokenized ECE implementation has been fully removed.

| Case | is Variable subscription | is virtual | Free Trial | Sign Up Fee | Expected | Actual | Observations |
| -- | -- | -- | -- | -- | -- | -- | -- |
| 1 | No | no | No | No | ✅ | ✅2 | No |
| 3 | No | no | Yes | No | ✅ | ✅ | Fixed. Automattic/woocommerce-payments#8152 closed. Free trial with deferred shipping now handled via subscription extensions in tokenized ECE.4 |
| 5 | no | yes | No | No | ✅ | ✅6 | no |
| 7 | no | yes | Yes | No | ❌ | ❌ | Intentionally blocked. Virtual free-trial subscriptions don't trigger address collection in ECE, so taxes can't be calculated. Blocked on product page by `is_product_supported()` and on cart/checkout by `is_cart_zero_total_with_trial_subscription()` (requires `needs_shipping()`).8 |
| 9 | yes | no | no | No | ✅ | ✅ |  |
| 10 | yes | no | no | Yes | ✅ | ✅ |  |
| 11 | yes | no | Yes | No | ✅ | ✅ | Fixed. Automattic/woocommerce-payments#8152 closed. Same deferred shipping support as case 3. |
| 12 | yes | no | Yes | Yes | ✅ | ✅ | Fixed. [#9776](<https://github.com/Automattic/woocommerce-payments/issues/9776>) closed, Automattic/woocommerce-payments#8152 closed. Same as case 4. Line item display still tracked by Automattic/woocommerce-payments#9799. |
| 13 | yes | Yes | No | No | ✅ | ✅ |  |
| 14 | yes | Yes | No | Yes | ✅ | ✅ |  |
| 15 | yes | Yes | Yes | No | ❌ | ❌ | Intentionally blocked. Same reason as case 7. Previously had incorrect amounts (fixed by [#9897](<https://github.com/Automattic/woocommerce-payments/issues/9897>) and [#9917](<https://github.com/Automattic/woocommerce-payments/issues/9917>)) but ECE is now disabled for virtual + trial combinations. |
| 16 | yes | Yes | Yes | Yes | ❌ | ❌ | Same as case 15/8. Blocked on product page. Cart/checkout conditional on tax settings. |

### Summary of changes since original matrix

**Resolved:**

* Cases 3, 4, 11, 12 (physical subscriptions with free trials) — now fully supported via tokenized ECE with deferred shipping (Automattic/woocommerce-payments#8152, [#9776](<https://github.com/Automattic/woocommerce-payments/issues/9776>), Automattic/woocommerce-payments#11340, Automattic/woocommerce-payments#11425)
* Cases 15, 16 amount display issues — fixed by [#9897](<https://github.com/Automattic/woocommerce-payments/issues/9897>) and [#9917](<https://github.com/Automattic/woocommerce-payments/issues/9917>) (but ECE is now intentionally disabled for these cases)
* Tokenized ECE is the default; old implementation removed

**Intentionally unsupported:**

* Cases 7, 8, 15, 16 (virtual + free trial) — ECE can't collect a shipping address for virtual products, making tax calculation inaccurate. These were previously marked as ✅ with incorrect amounts; they are now intentionally blocked.

**Remaining open item:**

* Automattic/woocommerce-payments#9799 — Line item details in the payment sheet for subscriptions with signup fees + free trials. The `wcpay.express-checkout.map-line-items` filter now shows recurring amounts with "(recurring)" labels and "First payment" dates, but Automattic/woocommerce-payments#9799 acceptance criteria (individual fee line items, explicit trial period display) may not be fully met.

```[tasklist]
### ECE Subscription Compatibility
- [x] https://github.com/Automattic/woocommerce-payments/issues/9771
- [x] https://github.com/Automattic/woocommerce-payments/issues/9776
- [x] https://github.com/Automattic/woocommerce-payments/issues/8152
- [ ] https://github.com/Automattic/woocommerce-payments/issues/9799
```

**Comment by asumaran** [comment]
Tokenized Carts for ECE allows us to retrieve the exact amounts the user will be charged, resolving many issues but also introducing new challenges. Since this new implementation for ECE [is about to be introduced](https://github.com/Automattic/woocommerce-payments/pull/10016), we will pause work on the ECE fixes for subscriptions to avoid making changes based on the soon-to-be deprecated implementation. We will wait for Tokenized ECE to be enabled by default.

I’ve tested all the cases related to this issue using #10099 with Tokenized Carts enabled. Which is the latest work related to this issue. And I can confirm that almost everything that was already working continues to work correctly with tokenized ECE, but I’ve noticed some new bugs:

For example, in cases 7, 12, and 15, ECE doesn’t work because the total amount to be paid is zero, and ECE doesn’t accept zero as a valid total. The payable amount is correct, but it can’t be used for ECE. I’m not sure how to proceed here.

In cases 4 and 12, we have another issue (Originally reported here paJDYF-g3R-p2#comment-27550)  The selected shipping option is not applied to the initial order, and the renewal order receives the first available shipping option, even though it was never selected. This is a bug.

In case 12 we also have another bug which is the same as @cesarcosta99 discovered [here](https://github.com/Automattic/woocommerce-payments/pull/9955#pullrequestreview-2507115855). ECE fails to open after closing the payment sheet. I plan to fix this as part of https://github.com/Automattic/woocommerce-payments/pull/10099

cc @bborman22 @pierorocca

**Comment by flakas** [comment]
Closing this epic — all four children are done:

- #9771 (Dec 2025), #9776 (Dec 2024), #8152 (Mar 2026), #9799 (Apr 2026)

**Comment by asumaran** [comment]
>all four children are done

I wouldn't consider them done. They were closed because there were no new reports, but the bugs probably still exist (I haven't worked on this in a long time), so I wouldn't consider them fixed.

**Comment by flakas** [comment]
Thanks for clarifying @asumaran!
I'm reviewing the backlog of all old tickets and either closing stale/completed tickets, or prepping/confirming tickets for the team to pick up part of the Woo Foundations work. Pardon me if i prematurely closed it, all related tickets seemed to be completed.

Looks like a couple of things landed after your last comment - WOOPMNT-5814 (ECE shipping for trial subscriptions, 10.7.0) and WOOPMNT-6081 (ECE pricing info for trials, 10.8.0), both this year.

Can we go off your January 2025 comment for what's left?

1. **Zero total carts** (cases 7, 12, 15) — ECE won't accept a zero total. At least partly handled since: #8152 covers zero-total trial carts on cart and checkout. The product page still hides ECE for a free trial with no signup fee, and your "not sure how to proceed" hasn't been answered.
2. **Shipping selection not applied** (cases 4, 12) — the selected option doesn't reach the initial order and the renewal gets the first available one. I can't find a ticket for this. WOOPMNT-5814 is close but fixed options not *showing*, not the selection being lost.
3. **ECE not reopening after the payment sheet closes** (case 12) — you were going to fix it in PR #10099, which ended up closed unmerged.

Do 2 and 3 still stand as far as you know? If so they're worth their own tickets — happy for you to file them since you have the case details, or I can try to write them up from your comment if you'd rather.
